lib/pbench/agent/tool_group.py

```diff
@@ -0,0 +1,158 @@
+import errno
+import os
+
+from pathlib import Path
+
+
+class BadToolGroup(Exception):
+    """Exception representing a tool group that does not exist or is invalid.
+    """
+
+    pass
+
+
+class ToolGroup:
+    """Provides an in-memory representation of the registered tools as recorded
+    on-disk.
+    """
+
+    # Current tool group prefix in use.
+    TOOL_GROUP_PREFIX = "tools-v1"
+
+    @staticmethod
+    def verify_tool_group(group, pbench_run=None):
+        """verify_tool_group - given a tool group name, verify it exists in the
+        ${pbench_run} directory as a properly prefixed tool group directory
+        name.
+
+        Raises a BadToolGroup exception if the directory is invalid or does not
+        exist, or if the pbench_run argument is None and the environment
+        variable of the same name is missing.
+
+        Returns a Pathlib object of the tool group directory on success.
+        """
+        _pbench_run = os.environ.get("pbench_run") if pbench_run is None else pbench_run
+        if not _pbench_run:
+            raise BadToolGroup(
+                "Cannot validate tool group, '{group}', 'pbench_run'"
+                " environment variable missing"
+            )
+
+        tg_dir_name = Path(_pbench_run, f"{ToolGroup.TOOL_GROUP_PREFIX}-{group}")
+        try:
+            tg_dir = tg_dir_name.resolve(strict=True)
+        except FileNotFoundError:
+            raise BadToolGroup(
+                f"Bad tool group, '{group}': directory {tg_dir_name} does not exist"
+            )
+        else:
+            if not tg_dir.is_dir():
+                raise BadToolGroup(
+                    f"Bad tool group, '{group}': directory {tg_dir_name} not valid"
+                )
+            else:
+                return tg_dir
+
+    def __init__(self, group):
+        """Construct a ToolGroup object from the on-disk data of the given
+        tool group.
+
+        If the given tool group is valid, the contents are read into the three
+        dictionary structures:
+
+          "toolnames" - each tool name is the key, with separate dictionaries
+          for each registered host
+
+          "hostnames" - each registered host is the key, with separate
+          dictionaries for each tool registered on that host
+
+          "labels" - each registered host name, that has a label, is the key,
+          and the label as the value; if a host is not labeled, it does not
+          show up in this dictionary
+
+        Raises BadToolGroup via the verify_tool_group() method on error.
+        """
+        self.tg_dir = self.verify_tool_group(group)
+        self.group = group
+
+        # __trigger__
+        try:
+            _trigger = (self.tg_dir / "__trigger__").read_text()
+        except OSError as ex:
+            if ex.errno != errno.ENOENT:
+                raise
+            # Ignore missing trigger file
+            self.trigger = None
```

Comment:
Could this be better expressed as the following?

```
        except FileNotFoundError:
            # Ignore missing trigger file
            self.trigger = None
```
Other `OSError`'s will simply propagate past.